<a href="https://colab.research.google.com/github/ezerwi/algo-quizzes/blob/master/py_0311.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### March/11/2020 Exercises

# BOJ/1546. 평균

In [0]:
n = int(input())

line = input().split()
nums = [int(i) for i in line]

# sum, maxi = 0,0
# for _ in nums :
#     if int(_) > maxi : maxi = int(_)
#     sum += int(_)
# print(maxi, sum)

# avg, sum2 = 0, 0
# for i in range(len(nums)) :
#     nums[i] = int(nums[i])/maxi * 100
#     sum2 += int(nums[i])
#     print(nums[i])

# avg = sum2/n
# print (sum2, avg)

maxi = int (max(nums))
sum1 = 0
for i in range(len(nums)) :
    sum1 += int(nums[i])
ret = float((sum1/maxi)/n*100)
print (ret)

3
40 80 60
75.0


### Short Coding

map 활용

map, lambda를 잘 쓰면 코드 길이가 확실히 크게 줄어드는 듯..

In [0]:
input()
*a,=map(int,input().split())
print(sum(a)*100/max(a)/len(a))

# BOJ/8958. OX퀴즈


In [0]:
for i in range(int(input())) :
    sum, temp = 0, 0
    for j in str(input()) :
        if j == 'X' :
            temp = 0
        elif j == 'O' :
            temp += 1
        sum += temp
    print (sum)


5
OOXXOXXOOO
10
OOXXOOXXOO
9
OXOXOXOXOXOXOX
7
OOOOOOOOOO
55
OOOOXOOOOXOOOOX
30


# BOJ/4344. 평균은 넘겠지

In [3]:
c = int(input())
exec("line = [ int(i) for i in input().split() ];avg = sum(line[1:])/line[0];over=[int(i) for i in line[1:] if i>avg];print('%0.3f%%' % (len(over)/int(line[0])*100) );"*c)


5
5 50 50 70 80 100
40.000%
7 100 95 90 80 70 60 50
57.143%
3 70 90 80
33.333%
3 70 90 81
66.667%
9 100 99 98 97 96 95 94 93 91
55.556%


# BOJ/15596. 정수 N개의 합

In [0]:
def solve(a) :
    return int(sum(a))

### Short Coding

이건 좀 거의 뭐 사기인듯..ㅋㅋ

In [0]:
solve=sum

# BOJ/4673. 셀프 넘버

초반에 문제 이해하는 것도 헷갈리고, 런타임 에러 세 번에

조건 하나를 잘못 생각해서 시간이 꽤 오래 걸림..

메모리		시간	    코드 길이

29436kb		84ms		266kb

In [0]:
# def sel(num) :
#     res = sum([int(i) for i in str(num)]+[int(num)])
#     return res
    
# number = [i for i in range(10001)]
# # nums = [1, 3, 5, 7, 9, 20, 31, 42, 53, 64, 75, 86, 97]

# for n in range(10001) :
#     # print (n, sel(n))
#     if sel(n)> 9999 : break
#     number[ sel(n) ] = 0

# res = sorted(set(number[:int(input())+1]))
# res.remove(0)
# for _ in res :
#     print(_)

In [0]:
def sel(num) :
    res = sum([int(i) for i in str(num)])+num
    return res
    
number = [i for i in range(10001)]

for n in range(10001) :
    if sel(n)> 10000 : continue
    print(n, sel(n))
    number[ sel(n) ] = 0

res = sorted(set(number))
res.remove(0)
for _ in res :
    print(_)

### Short Coding

map() 은 Iterator type..

python에서 sequence data type은 연산 가능..

In [0]:
r=range(9999);print(*sorted({*r}-{n+sum(map(int,str(n)))for n in r}))

In [51]:
print ( list ( map ( lambda x: str(x**2), range(1,15) )))
# print ( list ( map ( lambda s : int(s), input().split())) )

print ( sum(list(map ( lambda d : int(d), str(58329)))) + 58329 )

['1', '4', '9', '16', '25', '36', '49', '64', '81', '100', '121', '144', '169', '196']
53 34
[53, 34]
58356


# SWEA/D1/2072. 홀수만 더하기

53,596 kb
메모리
121 ms
실행시간
165
코드길이

In [0]:
for x in range(int(input())) :
	line, sum = list(map(lambda i : int(i), input().split())), 0
	for _ in line :
		if _%2==1 : sum+=_
	print ("#{} {}".format(x+1, sum))

# SWEA/D2/1859. 백만 장자 프로젝트

235,532 kb
메모리
1,442 ms
실행시간
288
코드길이

와 진짜 상상도 못할 데이터가 들어옴

원래 문제 자체도 군데군데 꼬아놔서 머리아픈데, test case 파일 자체가 무거워서 안열릴 정도면 말 다한 듯

In [0]:
# for i in range(int(input())) :
#     N = int(input())
#     days = list( map ( lambda i : int(i), input().split()))
#     res = 0
#     for t in range(N) :
#         if not days : break
#         m = max (days)
#         x = days.index(m)
#         del days[x]
#         res += sum ( [m-n for n in days[:x]] )
#         days = days[x:]
#     print(f'#{i+1} {res}')

In [0]:
for i in range(int(input())) :
    N = int(input())
    days = list( map ( lambda i : int(i), input().split()))
    res = 0
    max = days[-1]
    for t in range(N,0,-1) :
        if max > days[t-1] : res += max-days[t-1]
        else : max = days[t-1]
    print(f'#{i+1} {res}')

# SWEA/D2/1204. [S/W 문제해결 기본] 1일차 - 최빈수 구하기
간단해 보이는데 정답률이 생각보다 안높은데는 이유가 있었다

조건 확인 잘하자..

In [0]:
for _ in range(int(input())) :
    n = int(input())
    arr = [0 for i in range(101)]
    scores = list( map (lambda i : int(i), input().split()))
    for x in scores :
        arr[x] += 1
    
    print( f'#{n} {100-arr[::-1].index(max(arr))}' )